### PR TITLE
Sleep Analytics Backend - Part 1

### PR DESCRIPTION
**Pull request by jeffshek**

### CODE REVIEW

File: apis/betterself/v1/events/serializers.py [hunk 1]
```diff
@@ -215,3 +218,84 @@ def create(self, validated_data):
...
+    """
+    Custom serializer to parse sleep logs in a meaningful way
+
+    Returns a dataframe of sleep activity
+    """
+    def __init__(self, queryset):
+        self.sleep_activities = queryset
+
+    def get_sleep_history(self):
+        sleep_activities_values = self.sleep_activities.values('start_time', 'end_time')
```
Comment (jeffshek) [inline apis/betterself/v1/events/serializers.py]: Add timezone logic here and fix the fixtures generator to not be stupid

File: apis/betterself/v1/events/serializers.py [hunk 2]
```diff
@@ -215,3 +218,84 @@ def create(self, validated_data):
...
+    def __init__(self, queryset):
+        self.user_activities = queryset
+
+        try:
+            self.user = self.user_activities[0].user
+        except IndexError:
+            self.user = None
+
+    def get_user_activity_events(self):
+        activity_events_values = self.user_activities.values('time', 'user_activity__name')
```
Comment (jeffshek) [inline apis/betterself/v1/events/serializers.py]: if queryset is empty, return an empty dataframe? or fail earlier

File: apis/betterself/v1/events/serializers.py [hunk 3]
```diff
@@ -215,3 +218,84 @@ def create(self, validated_data):
...
+        self.sleep_activities = queryset
+
+    def get_sleep_history(self):
+        sleep_activities_values = self.sleep_activities.values('start_time', 'end_time')
+
+        # for each given 24 hour period (ending at 2PM - the latest I can imagine someone might sleep in)
+        # lot of mental debate here between calculating the sleep one gets from monday 10PM to tuesday 6AM as
+        # either a Monday or Tuesday night, but I've decided to lean toward calculating that as Monday night
+        dataframe = pd.DataFrame.from_records(sleep_activities_values)
+        dataframe['sleep_time'] = dataframe['end_time'] - dataframe['start_time']
```
Comment (jeffshek) [inline apis/betterself/v1/events/serializers.py]: Should you change this to a constant?

File: apis/betterself/v1/events/serializers.py [hunk 4]
```diff
@@ -215,3 +218,84 @@ def create(self, validated_data):
...
+def round_timestamp_to_sleep_date(timeseries):
+    """
+    Not my proudest function ... this isn't as efficient as it could be, but struggling
+    with some pandas syntax to find the perfect pandas one-line
+
+    This can be much more performant, but need time to sit down and figure it out
+    """
+    sleep_dates = []
+    for value in timeseries:
+        if value.hour < 5:
```
Comment (jeffshek) [inline apis/betterself/v1/events/serializers.py]: 5 should be a change in settings somewhere ...

File: apis/betterself/v1/events/tests/test_sleep_views.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+from rest_framework.test import APIClient
+
+from apis.betterself.v1.signup.fixtures.builders import DemoHistoricalDataBuilder
+
+# python manage.py test apis.betterself.v1.events.tests.test_sleep_views
+from constants import SLEEP_MINUTES_COLUMN
+from events.models import SleepActivity
+
+User = get_user_model()
+
```
Comment (jeffshek) [inline apis/betterself/v1/events/tests/test_sleep_views.py]: Add tests with empty users! See what happens to the views

File: apis/betterself/v1/events/views.py [hunk 6]
```diff
@@ -85,3 +92,70 @@ def get_serializer_class(self):
...
+
+class SleepAggregatesView(APIView):
+    def get(self, request):
+        user = request.user
+        sleep_activities = SleepActivity.objects.filter(user=user)
+
+        serializer = SleepActivityDataframeBuilder(sleep_activities)
+        sleep_aggregate = serializer.get_sleep_history()
+
+        # this is crap, there's got to be something you're missing with pandas and drf json responses
```
Comment (jeffshek) [inline apis/betterself/v1/events/views.py]: Explain what's actually going on

File: apis/betterself/v1/events/views.py [hunk 7]
```diff
@@ -85,3 +92,70 @@ def get_serializer_class(self):
...
+        return Response(data=result, content_type='application/json')
+
+
+class SleepAveragesView(APIView):
+    def get(self, request):
+        try:
+            lookback = int(request.query_params[LOOKBACK_PARAM_NAME])
+        except (ValueError, MultiValueDictKeyError):
+            # MultiValueDictKeyError when a key doesn't exist
+            # if something that was entered for a lookback that couldn't be interpreted
```
Comment (jeffshek) [inline apis/betterself/v1/events/views.py]: write ValueError

File: apis/betterself/v1/events/views.py [hunk 8]
```diff
@@ -85,3 +92,70 @@ def get_serializer_class(self):
...
+        return Response(data=result, content_type='application/json')
+
+
+class SleepActivitiesCorrelationView(APIView):
+    def get(self, request):
+        user = request.user
+
+        sleep_activities = SleepActivity.objects.filter(user=user)
+        sleep_serializer = SleepActivityDataframeBuilder(sleep_activities)
+        # have to clean this up a lot ...
```
Comment (jeffshek) [inline apis/betterself/v1/events/views.py]: Do you still have to worry about timezones here?

File: apis/betterself/v1/events/views.py [hunk 9]
```diff
@@ -85,3 +92,70 @@ def get_serializer_class(self):
...
+        sleep_aggregate = sleep_serializer.get_sleep_history()
+        sleep_aggregate = sleep_aggregate.tz_convert(user.pytz_timezone)
+
+        # resample so that it goes from no frequency to a daily frequency
+        # which matches UserActivityEvents, eventually need to be more elegant
+        sleep_aggregate = sleep_aggregate.resample('d').last()
```
Comment (jeffshek) [inline apis/betterself/v1/events/views.py]: Make this a capital D for OCD

Comment (jeffshek) [inline apis/betterself/v1/events/views.py]: sum() should result in the same, do it just in case.

File: apis/betterself/v1/events/serializers.py (hunk 4 above)
Comment (jeffshek) [inline apis/betterself/v1/events/serializers.py]: Eventually on the user settings level, but make it a user constant for now.
